src/datahike/schema.cljc

```diff
@@ -0,0 +1,136 @@
+                                                  :db/unique :db.unique/identity
+                                                  :db/index true
+                                                  :db/cardinality :db.cardinality/one}
+                                   :db/retracted {:db/valueType :db.type/long
+                                                  :db/unique :db.unique/identity
+                                                  :db/cardinality :db.cardinality/many}
+                                   :db/doc {:db/valueType :db.type/string
+                                            :db/index true
+                                            :db/cardinality :db.cardinality/one}
+                                   :db.install/_attribute {:db/valueType   :db.type.install/_attribute
+                                                           :db/unique      :db.unique/identity
+                                                           :db/cardinality :db.cardinality/one}})
+
+(def schema-keys #{:db/ident :db/isComponent :db/valueType :db/cardinality :db/unique :db/index :db.install/_attribute :db/doc})
+(def meta-keys #{:db/txInstant :db/retracted})
+
+(defn meta-attr? [attr]
+  (s/valid? ::meta-attribute attr))
+
+(defn meta-entity? [entity]
+  (some #(contains? entity %) meta-keys))
+
+(defn schema-attr? [attr]
+  (s/valid? ::schema-attribute attr))
+
+(defn value-valid? [[_ _ a v _] schema]
+  (let [schema (if (or (meta-attr? a) (schema-attr? a))
+                 implicit-schema-spec
+                 schema)
+        value-type (get-in schema [a :db/valueType])]
+    (s/valid? value-type v)))
+
+(defn instant? [^Datom datom schema]
+  (let [a (.-a datom)
+        schema (if (or (meta-attr? a) (schema-attr? a))
+                 implicit-schema-spec
+                 schema)]
+    (= (get-in schema [a :db/valueType]) :db.type/instant)))
+
+(defn schema-entity? [entity]
+  (some #(contains? entity %) schema-keys))
+
+(defn schema? [schema]
+  (s/valid? ::schema schema))
+
+(defn describe-type [schema-type]
+  (s/describe schema-type))
+
+(defn find-invalid-schema-updates [entity attr-schema]
+  (reduce-kv
+   (fn [m attr-def new-value]
+     (let [old-value (get-in attr-schema [attr-def])]
+       (when-not (= old-value new-value)
+         (case attr-def
+           :db/cardinality (if (= new-value :db.cardinality/many)
+                             (if (get-in attr-schema [:db/unique])
+                               (assoc m attr-def [old-value new-value])
+                               nil)
+                             (assoc m attr-def [old-value new-value]))
+           :db/unique (when-not (get-in attr-schema [:db/unique])
+                        (when-not (= (get-in attr-schema [:db/cardinality]) :db.cardinality/one)
+                          (assoc m attr-def [old-value new-value])))
+           (assoc m attr-def [old-value new-value]))))) {} (dissoc entity :db/id)))
```

Comment:
I would line break here in front of `{}` to make the arguments of `reduce-kv` more easily visible.

Reply:
Fixed with #70